File: app/src/test/java/io/github/h3nb/jlmodplus/filepicker/FilePickerModelTest.java
/*
 * Copyright 2026 H3NB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.h3nb.jlmodplus.filepicker;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FilePickerModelTest {
	@Rule
	public final TemporaryFolder temporaryFolder = new TemporaryFolder();

	@Test
	public void modeValuesRemainCompatibleWithTheLegacyContract() {
		assertEquals(0, FilePickerContract.MODE_FILE);
		assertEquals(1, FilePickerContract.MODE_DIR);
		assertEquals(2, FilePickerContract.MODE_FILE_AND_DIR);
		assertEquals(3, FilePickerContract.MODE_NEW_FILE);
	}

	@Test
	public void selectedFileRestoresItsParentDirectory() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File games = new File(root, "games");
		assertTrue(games.mkdir());
		File selected = new File(games, "Game.JAR");
		assertTrue(selected.createNewFile());

		assertEquals(
				FilePickerModel.canonicalFile(games),
				FilePickerModel.normalizeStartPath(selected, root));
	}

	@Test
	public void missingPathFallsBackToExistingParentOrRoot() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File existingParent = new File(root, "games");
		assertTrue(existingParent.mkdir());

		assertEquals(
				FilePickerModel.canonicalFile(existingParent),
				FilePickerModel.normalizeStartPath(new File(existingParent, "new.jar"), root));
		assertEquals(
				FilePickerModel.canonicalFile(root),
				FilePickerModel.normalizeStartPath(new File(root, "missing/new.jar"), root));
	}

	@Test
	public void canonicalPathMustRemainInsidePickerRoot() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File child = new File(root, "games");
		assertTrue(child.mkdir());

		assertTrue(FilePickerModel.isWithinRoot(child, root));
		assertFalse(FilePickerModel.isWithinRoot(new File(root, "../outside"), root));
		assertEquals(
				FilePickerModel.canonicalFile(root),
				FilePickerModel.normalizeStartPath(new File(root, "../outside"), root));
	}

	@Test
	public void extensionFilterIsCaseInsensitiveAndModeAware() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File jar = new File(root, "GAME.JAR");
		File txt = new File(root, "notes.txt");
		assertTrue(jar.createNewFile());
		assertTrue(txt.createNewFile());

		assertTrue(FilePickerModel.isItemVisible(jar, FilePickerContract.MODE_FILE));
		assertFalse(FilePickerModel.isItemVisible(txt, FilePickerContract.MODE_FILE));
		assertFalse(FilePickerModel.isItemVisible(jar, FilePickerContract.MODE_DIR));
	}

	@Test
	public void directoriesSortBeforeFilesWithDeterministicTieBreakers() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File zFile = new File(root, "z.jar");
		File aDirectory = new File(root, "A");
		File bDirectory = new File(root, "b");
		assertTrue(zFile.createNewFile());
		assertTrue(aDirectory.mkdir());
		assertTrue(bDirectory.mkdir());

		List<File> files = new ArrayList<>(Arrays.asList(zFile, bDirectory, aDirectory));
		FilePickerModel.sortFiles(files);

		assertEquals(aDirectory, files.get(0));
		assertEquals(bDirectory, files.get(1));
		assertEquals(zFile, files.get(2));
	}

	@Test
	public void directoryNameValidationRejectsPathsAndAmbiguousNames() {
		assertTrue(FilePickerModel.isValidDirectoryName("Games 2026"));
		assertFalse(FilePickerModel.isValidDirectoryName(null));
		assertFalse(FilePickerModel.isValidDirectoryName(""));
		assertFalse(FilePickerModel.isValidDirectoryName(" . "));
		assertFalse(FilePickerModel.isValidDirectoryName("."));
		assertFalse(FilePickerModel.isValidDirectoryName(".."));
		assertFalse(FilePickerModel.isValidDirectoryName("a/b"));
		assertFalse(FilePickerModel.isValidDirectoryName("a\\b"));
	}

	@Test
	public void selectionRulesMatchFileAndDirectoryModes() throws IOException {
		File root = temporaryFolder.newFolder("storage");
		File directory = new File(root, "games");
		File jar = new File(root, "game.jar");
		assertTrue(directory.mkdir());
		assertTrue(jar.createNewFile());

		assertTrue(FilePickerModel.isSelectable(directory, FilePickerContract.MODE_DIR, false));
		assertFalse(FilePickerModel.isSelectable(directory, FilePickerContract.MODE_FILE, false));
		assertTrue(FilePickerModel.isSelectable(jar, FilePickerContract.MODE_FILE, false));
		assertTrue(FilePickerModel.isSelectable(jar, FilePickerContract.MODE_FILE_AND_DIR, false));
	}
}

File: app/src/test/java/ru/playsoftware/j2meloader/filepicker/FilePickerModelTest.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ru.playsoftware.j2meloader.filepicker

import java.io.File
import java.nio.file.Files
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class FilePickerModelTest {
    @Test
    fun filtersOnlySupportedFilesAndSortsFoldersFirst() {
        val root = Files.createTempDirectory("jlmod-picker").toFile()
        try {
            File(root, "zeta.jar").createNewFile()
            File(root, "alpha.JAD").createNewFile()
            File(root, "ignored.txt").createNewFile()
            File(root, ".hidden.jar").createNewFile()
            File(root, "Games").mkdir()

            val entries = FilePickerRules.sortEntries(
                root.listFiles()!!.asIterable(),
                FilePickerContract.MODE_FILE,
            )

            assertEquals(listOf("Games", "alpha.JAD", "zeta.jar"), entries.map { it.name })
            assertTrue(entries.first().kind == FilePickerEntryKind.DIRECTORY)
        } finally {
            root.deleteRecursively()
        }
    }

    @Test
    fun searchAndSortAreIndependentOfFilesystemImplementation() {
        val entries = listOf(
            FilePickerEntry("/a", "Zeta.jar", FilePickerEntryKind.FILE, modifiedAt = 3L),
            FilePickerEntry("/b", "alpha.jar", FilePickerEntryKind.FILE, modifiedAt = 1L),
            FilePickerEntry("/c", "Games", FilePickerEntryKind.DIRECTORY, modifiedAt = 2L),
        )

        assertEquals(
            listOf("alpha.jar", "Zeta.jar"),
            FilePickerRules.filterAndSort(
                entries,
                query = ".jar",
                sortOrder = FilePickerSortOrder.NAME_ASCENDING,
            ).map { it.name },
        )
        assertEquals(
            listOf("Zeta.jar", "Games", "alpha.jar"),
            FilePickerRules.filterAndSort(
                entries,
                query = "",
                sortOrder = FilePickerSortOrder.MODIFIED_NEWEST,
            ).map { it.name },
        )
    }

    @Test
    fun modeContractsKeepDirectoryAndExistingFileSelectionDistinct() {
        val existingFile = Files.createTempFile("existing", ".jar").toFile()
        try {
            val directoryRequest = FilePickerRequest(
                startPath = null,
                mode = FilePickerContract.MODE_FILE_AND_DIR,
                allowMultiple = false,
                singleClick = false,
                allowCreateDirectory = false,
                allowExistingFile = true,
            )
            val multipleRequest = directoryRequest.copy(allowMultiple = true)
            val directoryState = FilePickerState(
                request = directoryRequest,
                rootPath = "/storage",
                currentPath = "/storage/emulated/0",
            )

            assertTrue(directoryState.canConfirm)
            assertFalse(directoryState.copy(request = multipleRequest).canConfirm)
            assertTrue(
                FilePickerRules.isVisible(
                    existingFile,
                    FilePickerContract.MODE_NEW_FILE,
                    allowExistingFile = true,
                ),
            )
            assertFalse(
                FilePickerRules.isVisible(
                    existingFile,
                    FilePickerContract.MODE_NEW_FILE,
                    allowExistingFile = false,
                ),
            )
        } finally {
            existingFile.delete()
        }
    }

    @Test
    fun startPathAndParentStayInsideRoot() {
        val root = Files.createTempDirectory("jlmod-root").toFile()
        val nested = File(root, "emulated/0").apply { mkdirs() }
        try {
            assertEquals(nested, FilePickerRules.normalizeStartPath(nested.path, root))
            assertEquals(root, FilePickerRules.normalizeStartPath(File(root, "outside").path, root))
            assertEquals(
                root,
                FilePickerRules.normalizeStartPath(File(root, "../outside").path, root),
            )
            assertEquals(File(root, "emulated"), FilePickerRules.parent(nested, root))
            assertFalse(FilePickerRules.isValidFolderName("../escape"))
            assertTrue(FilePickerRules.isValidFolderName("My Games"))
        } finally {
            root.deleteRecursively()
        }
    }
}
